plugin/backend/ipvariant.d

```diff
@@ -0,0 +1,906 @@
+            auto visitor = scoped!ClassVisitor(v, ns_stack, container, indent + 1);
+            v.accept(visitor);
+
+            root.put(visitor.root);
+            container.put(visitor.root, visitor.root.fullyQualifiedName);
+        } else {
+            auto type = retrieveType(v.cursor, container, indent);
+            put(type, container, indent);
+        }
+    }
+
+    override void visit(const(Namespace) v) @trusted {
+        mixin(mixinNodeLog!());
+
+        () @trusted{ ns_stack ~= CppNs(v.cursor.spelling); }();
+        // pop the stack when done
+        scope (exit)
+            ns_stack = ns_stack[0 .. $ - 1];
+
+        auto ns_visitor = scoped!(CppVisitor!(CppNamespace, ControllerT, ProductT))(ctrl,
+                prod, indent, ns_stack, container);
+
+        // fill the namespace with content from the analyse
+        v.accept(ns_visitor);
+        root.put(ns_visitor.root);
+    }
+
+    override void visit(const(TranslationUnit) v) {
+        import std.algorithm : filter;
+        import cpptooling.analyzer.clang.type : makeLocation;
+
+        mixin(mixinNodeLog!());
+
+        LocationTag tu_loc;
+        () @trusted{ tu_loc = LocationTag(Location(v.cursor.spelling, 0, 0)); }();
+
+        if (tu_loc.kind != LocationTag.Kind.noloc && ctrl.doFile(tu_loc.file,
+                "root " ~ tu_loc.toString)) {
+            prod.putLocation(FileName(tu_loc.file), LocationType.Root);
+        }
+
+        v.accept(this);
+    }
+
+    void toString(Writer)(scope Writer w) @safe const {
+        import std.format : FormatSpec;
+        import std.range.primitives : put;
+
+        auto fmt = FormatSpec!char("%u");
+        fmt.writeUpToNextSpec(w);
+
+        root.toString(w, fmt);
+        put(w, "\n");
+        container.get.toString(w, FormatSpec!char("%s"));
+    }
+
+    override string toString() const {
+        import std.exception : assumeUnique;
+
+        char[] buf;
+        buf.reserve(100);
+        toString((const(char)[] s) { buf ~= s; });
+        auto trustedUnique(T)(T t) @trusted {
+            return assumeUnique(t);
+        }
+
+        return trustedUnique(buf);
+    }
+}
+
+private:
+@safe:
+
+import cpptooling.data.representation : CppRoot, CppClass, CppMethod, CppCtor,
+    CppDtor, CFunction, CppNamespace;
+import cpptooling.data.type : LocationTag, Location;
+import cpptooling.data.symbol.container : Container;
+import dsrcgen.cpp : E;
+
+enum ClassType {
+    Normal,
+    Adapter,
+    Gmock
+}
+
+enum NamespaceType {
+    Normal,
+    TestDoubleSingleton,
+    TestDouble
+}
+
+/** Structurally transform the input to a stub implementation.
+ *
+ * TODO should handle StorageClass like cvariant do.
+ *
+ * Ignoring C functions and globals by ignoring the root ranges funcRange and
+ * globalRange.
+ *
+ * Params:
+ *  ctrl: removes according to directives via ctrl
+ */
+CppT rawFilter(CppT, LookupT)(CppT input, Controller ctrl, Products prod, LookupT lookup) @safe {
```

Comment:
consider the design of the data flow throughout ipvariant.
What is rawFilter responsible for?
What is translate responsible for?
Does the current implementation meet the design?

See (1) for additional thoughts